wasm_node_api/wasm_node_api.cc

```diff
@@ -0,0 +1,886 @@
+#include <type_traits>
+#include <functional>
+#include <uv.h>
+
+#define NODE_WANT_INTERNALS 1
+#include "js_native_api_v8.h"
+#include "wasm_node_api.h"
+#include "node_api.h"
+#include "v8-wasm.h"
+#include "env.h"
+#undef NODE_WANT_INTERNALS
+
+namespace node {
+namespace wasm_napi {
+
+namespace w = v8::wasm;
+
+using callback = w::Func::callbackType;
+
+template<typename T>
+struct is_function_pointer {
+  static const bool value =
+    std::is_pointer<T>::value ?
+      std::is_function<typename std::remove_pointer<T>::type>::value :
+      false;
+};
+
+template<typename T>
+constexpr auto is_function_pointer_v = is_function_pointer<T>::value;
+
+template<typename T>
+class Native {
+public:
+  using type = T;
+};
+
+template<typename T>
+class Wasm {
+public:
+  using type = T;
+};
+
+template<typename>
+struct is_native : public std::false_type {};
+
+template<typename T>
+struct is_native<Native<T>> : public std::true_type {};
+
+template <typename T>
+constexpr auto is_native_v = is_native<T>::value;
+
+template<typename>
+struct is_wasm : public std::false_type {};
+
+template<typename T>
+struct is_wasm<Wasm<T>> : public std::true_type {};
+
+template <typename T>
+constexpr auto is_wasm_v = is_wasm<T>::value;
+
+template<class K>
+using kind = typename K::type;
+
+template<typename T>
+constexpr w::ValKind wasmType() {
+  if constexpr (std::is_same<kind<T>, float>::value) {
+    return w::ValKind::F32;
+  } else if constexpr (std::is_same<kind<T>, double>::value) {
+    return w::ValKind::F64;
+  #if __WORDSIZE == 64
+  } else if constexpr (is_native_v<T> && std::is_pointer<kind<T>>::value) {
+    return w::ValKind::I64;
+  #endif
+  } else {
+    return w::ValKind::I32;
+  }
+}
+
+template<w::ValKind... Args>
+constexpr auto exportVoid(callback cb) -> w::Func {
```

Comment:
These two functions (`exportVoid` and `exportFn`) allow us to turn a given WASM callback into a `v8::wasm::Func` value (that can be shared with WASM code). It does this by constructing the appropriate `v8::wasm::FuncType` based on the type parameters passed in and creating a `v8::wasm::Func` from the callback.

Reply:
This should be a comment in the code.